packages/core/mesh/edge-client/src/edge-client.ts

```diff
@@ -122,97 +137,96 @@ export class EdgeClient extends Resource implements EdgeConnection {
+
+    const identity = this._identity;
+    const path = `/ws/${identity.identityKey}/${identity.peerKey}`;
     const protocolHeader = this._config.disableAuth ? undefined : await this._createAuthHeader(path);
+    if (this._identity !== identity) {
+      log('identity changed during auth header request');
+      return undefined;
+    }
 
+    const restartRequired = new Trigger();
     const url = new URL(path, this._baseWsUrl);
     log('Opening websocket', { url: url.toString(), protocolHeader });
-    this._ws = new WebSocket(url, protocolHeader ? [protocolHeader] : []);
-
-    this._ws.onopen = () => {
-      log('opened', this.info);
-      this._ready.wake();
-      this.connected.emit();
-    };
-    this._ws.onclose = () => {
-      log('closed', this.info);
-      this._persistentLifecycle.scheduleRestart();
-    };
-    this._ws.onerror = (event) => {
-      log.warn('EdgeClient socket error', { error: event.error, info: event.message });
-      this._persistentLifecycle.scheduleRestart();
-    };
-    /**
-     * https://developer.mozilla.org/en-US/docs/Web/API/MessageEvent/data
-     */
-    this._ws.onmessage = async (event) => {
-      if (event.data === '__pong__') {
-        this._onHeartbeat();
-        return;
-      }
-      const data = await toUint8Array(event.data);
-      const message = buf.fromBinary(MessageSchema, data);
-      log('received', { peerKey: this._identity.peerKey, payload: protocol.getPayloadType(message) });
-      if (message) {
-        for (const listener of this._listeners) {
-          try {
-            await listener(message);
-          } catch (err) {
-            log.error('processing', { err, payload: protocol.getPayloadType(message) });
+    const connection = new EdgeWsConnection(
+      identity,
+      { url, protocolHeader },
+      {
+        onConnected: () => {
+          if (this._isActive(connection)) {
+            this._ready.wake();
+            this._notifyReconnected();
+          } else {
+            log.verbose('connected callback ignored, because connection is not active');
           }
-        }
-      }
-    };
-
-    // TODO(dmaretskyi): Potential race condition here since web socket errors don't resolve this trigger.
-    await this._ready.wait({ timeout: this._config.timeout ?? DEFAULT_TIMEOUT });
-    log('Websocket is ready', { identity: this._identity.identityKey, peer: this._identity.peerKey });
-
-    // TODO(dmaretskyi): Potential leak: context re-assigned without disposing the previous one.
-    this._keepaliveCtx = new Context();
-    scheduleTaskInterval(
-      this._keepaliveCtx,
-      async () => {
-        // TODO(mykola): use RFC6455 ping/pong once implemented in the browser?
-        // Cloudflare's worker responds to this `without interrupting hibernation`. https://developers.cloudflare.com/durable-objects/api/websockets/#setwebsocketautoresponse
-        this._ws?.send('__ping__');
+        },
+        onRestartRequired: () => {
```

Comment:
`onClose` is better I think -- callback names should descrive the event happening in the class defining the callback, not what the implementation does